### PR TITLE
Run unit tests on Windows as part of CI build

### PR DESCRIPTION
**Pull request by georgeharley**
* WIP
* This is connected to #1001

### CODE REVIEW

**Comment by fbiville** [comment]
`choco install make` works, that's awesome! It probably means https://travis-ci.community/t/choco-install-hangs-forever/307/6 is no longer relevant.

File: .travis.yml [hunk 1]
```diff
@@ -10,10 +10,22 @@ install: true
 matrix:
   include:
   - stage: test
+    name: "Test on Linux"
```
Comment (ericbottard) [inline .travis.yml]: Shouldn't that require `os: linux` for 100% symmetry?

File: Makefile [hunk 2]
```diff
@@ -1,15 +1,35 @@
 .PHONY: build clean test all release gen-mocks check-mockery verify-docs clean-docs
 
-OUTPUT = ./riff
-GO_SOURCES = $(shell find . -type f -name '*.go')
-VERSION ?= $(shell cat VERSION)
+ISWIN:=
```
Comment (ericbottard) [inline Makefile]: Can this be removed?

File: Makefile [hunk 3]
```diff
@@ -1,15 +1,35 @@
...
+ifeq ($(OS),Windows_NT)
+	ISWIN=true
+    OUTPUT = .\riff
```
Comment (ericbottard) [inline Makefile]: Can't we merge this AND the extension in OUTPUT?

File: Makefile [hunk 4]
```diff
@@ -1,15 +1,35 @@
...
+    OUTPUT_EXT = .exe
+	VERSION = $(shell type VERSION)
+	GITDIRTY = $(shell git diff --quiet HEAD || echo dirty)
```
Comment (ericbottard) [inline Makefile]: I think the quotes in the *nix version are not necessary, so could be factored out

File: Makefile [hunk 5]
```diff
@@ -1,15 +1,35 @@
...
+	GOBIN = $(shell go env GOPATH)\bin
+	MOCKERY_CHECK_CMD = where mockery > NUL || (echo mockery not found: issue "go get -u github.com/vektra/mockery/.../" && exit 1)
+	COPY_CMD = copy
+	FILE_DELETE_CMD = del /f /q 2>NUL
+	TOUCH_FILE_CMD = copy /y nul
+else
+	ISWIN=false
+	OUTPUT = ./riff
+    OUTPUT_EXT =
+	VERSION ?= $(shell cat VERSION)
```
Comment (ericbottard) [inline Makefile]: Why the `?=` here and not there?

File: Makefile [hunk 6]
```diff
@@ -37,34 +57,45 @@ gen-mocks: check-mockery
...
-	cp $(OUTPUT) $(GOBIN)
+	$(COPY_CMD) $(OUTPUT)$(OUTPUT_EXT) $(GOBIN)
 
-$(OUTPUT): $(GO_SOURCES) vendor VERSION
-	go build -o $(OUTPUT) -ldflags "$(LDFLAGS_VERSION)"
+$(OUTPUT): vendor VERSION
+	go build -o $(OUTPUT)$(OUTPUT_EXT) -ldflags "$(LDFLAGS_VERSION)"
 
-release: $(GO_SOURCES) vendor VERSION
+release: vendor VERSION
```
Comment (ericbottard) [inline Makefile]: I don't think we need to bother performing the release ON windows

File: Makefile [hunk 7]
```diff
@@ -37,34 +57,45 @@ gen-mocks: check-mockery
...
+	cmd /C "set GOOS=darwin&&set GOARCH=amd64&& go build -ldflags "$(LDFLAGS_VERSION)" -o $(OUTPUT) && 7z a -ttar riff-darwin-amd64.tar $(OUTPUT) && 7z a -tzip riff-darwin-amd64.tgz riff-darwin-amd64.tar && del /q riff-darwin-amd64.tar && del /q $(OUTPUT)"
+	cmd /C "set GOOS=linux&&set GOARCH=amd64&& go build -ldflags "$(LDFLAGS_VERSION)" -o $(OUTPUT) && 7z a -ttar riff-linux-amd64.tar $(OUTPUT) && 7z a -tzip riff-linux-amd64.tgz riff-linux-amd64.tar && del /q riff-linux-amd64.tar && del /q $(OUTPUT)"
+else	
 	GOOS=darwin   GOARCH=amd64 go build -ldflags "$(LDFLAGS_VERSION)" -o $(OUTPUT)     && tar -czf riff-darwin-amd64.tgz $(OUTPUT) && rm -f $(OUTPUT)
 	GOOS=linux    GOARCH=amd64 go build -ldflags "$(LDFLAGS_VERSION)" -o $(OUTPUT)     && tar -czf riff-linux-amd64.tgz $(OUTPUT) && rm -f $(OUTPUT)
-	GOOS=windows  GOARCH=amd64 go build -ldflags "$(LDFLAGS_VERSION)" -o $(OUTPUT).exe && zip -mq riff-windows-amd64.zip $(OUTPUT).exe && rm -f $(OUTPUT).exe
+	GOOS=windows  GOARCH=amd64 go build -ldflags "$(LDFLAGS_VERSION)" -o $(OUTPUT)$(OUTPUT_EXT) && zip -mq riff-windows-amd64.zip $(OUTPUT)$(OUTPUT_EXT) && rm -f $(OUTPUT)$(OUTPUT_EXT)
+endif
 
 docs: $(OUTPUT) clean-docs
```
Comment (ericbottard) [inline Makefile]: This is why it's better to merge OUTPUT and OUTPUT_EXT: this is a *dependency*, needs to look at the actual produced file mod date

File: .travis.yml (hunk 1 above)
Comment (georgeharley) [reply .travis.yml]: Yes, good point. Added now. Thanks.

**Comment by georgeharley** [comment]
@ericbottard aftr discovering that Travis Windows jobs run in a Git Bash shell I've drastically cut down the number of code changes. It's now just a matter of updating the Travis configuration to run the new Windows  job.

File: .travis.yml [hunk 8]
```diff
@@ -10,10 +10,20 @@ install: true
...
     - "go get -u github.com/vektra/mockery/.../"
     script: make build test verify-docs
+  - stage: test
+    name: "Test on Windows (Git Bash)"
+    os: windows
+    install:
+    - "choco install make"
+    - "go get -u github.com/golang/dep/cmd/dep"
+    - "go get -u github.com/vektra/mockery/.../"
+    script: make build test verify-docs
```
Comment (ericbottard) [inline .travis.yml]: I thought there was a problem with deleting files on windows+choco?

Comment (georgeharley) [reply .travis.yml]: @ericbottard Yes, there was a problem with trying to delete files using the Windows `rmdir` command but investigating that led to the realisation that Travis builds in Windows environments actually run in the Git Bash shell and the Makefile was actually invoking the `rmdir` GNU utility. Travis' choice of Git Bash means our Makefile works as-is (and hence the removal of all the WIP changes to it).

Comment (ericbottard) [reply .travis.yml]: Excellent!

**Pull request merged**